### PR TITLE
Use Karma/BrowserStack to run <model-viewer> tests

### PR DESCRIPTION
**Pull request by cdata**
This change grew substantially out of scope, but one thing led to another and now our test toolchain and CI pipeline is totally different ✨

This change updates our pipeline to run in Github Actions instead of Travis. This is good because it brings us closer to Github's well-tread cowpaths, and moves us away from Travis which seems to be more-or-less frozen in time feature/support-wise (probably because of [their recent sale](https://news.ycombinator.com/item?id=18978251)).

Secondly, this change moves us from SauceLabs to BrowserStack for our x-browser testing needs. This has generally yielded an increase in overall performance for our x-browser tests, as well as improved reliability for iOS Safari tests and unlocks the potential to re-introduce our Android Chrome tests in the future. We might even get crazy and add tests for Samsung Internet someday. Future work on this tracked in #691 .

Thirdly, this change moves our test toolchain from WCT (which is [deprecated](https://github.com/Polymer/tools/issues/3493)) to Karma (which has an uncertain future, but is way better than WCT and [has active support for now](https://github.com/karma-runner/karma/pulse/monthly)). Thankfully, only minimal changes were required to our test suites to support this shift. Most of the time spent on this change went to ensuring that the combination of Karma, BrowserStack and Github Actions all worked in harmony and as intended.

And the coup de grâce: `npm run dev` does something useful for the `@google/model-viewer` package now (more or less what it does for `@google/3dom`), so running all the build tools simultaneously in a single terminal just got easier.

A final note on Travis: this PR continues to run Travis builds even though the Travis config has been removed. Travis has to be manually turned off for our repository before those status lines will go away. I will do that after this PR is merged.

An example 3P PR has been set up here to demonstrate how the build changes: https://github.com/google/model-viewer/pull/1082

### CODE REVIEW

File: .github/workflows/unit-tests-legacy.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    - name: Generate legacy bundles
+      run: npm run build:legacy-support
+
+    - name: Unit tests
+      env:
+        BROWSER_STACK_USERNAME: ${{ secrets.BROWSER_STACK_USERNAME }}
+        BROWSER_STACK_ACCESS_KEY: ${{ secrets.BROWSER_STACK_ACCESS_KEY }}
+      run: |
+        # Don't run Sauce from a forked repo, since the code may not be trusted
```
Comment (elalish) [inline .github/workflows/unit-tests-legacy.yml]: If this is still true, we should probably at least update that it's not Sauce anymore.

File: .github/workflows/unit-tests.yml [hunk 2]
```diff
@@ -39,15 +39,13 @@ jobs:
 
     - name: Unit tests
       env:
-        SAUCE_USERNAME: ${{ secrets.SAUCE_USERNAME }}
-        SAUCE_ACCESS_KEY: ${{ secrets.SAUCE_ACCESS_KEY }}
+        BROWSER_STACK_USERNAME: ${{ secrets.BROWSER_STACK_USERNAME }}
+        BROWSER_STACK_ACCESS_KEY: ${{ secrets.BROWSER_STACK_ACCESS_KEY }}
       run: |
         # Don't run Sauce from a forked repo, since the code may not be trusted
```
Comment (elalish) [inline .github/workflows/unit-tests.yml]: Same as above; maybe just do a search for 'Sauce'.

File: .travis.yml [hunk 3]
```diff
@@ -26,7 +26,7 @@ install:
 - npm run bootstrap
 - npm run build
 script:
-- npm run test:ci
+- npm run wct-local-chrome
```
Comment (elalish) [inline .travis.yml]: WCT is gone now, right?

File: packages/3dom/karma.conf.js [hunk 4]
```diff
@@ -60,65 +60,117 @@ module.exports = function(config) {
...
+        const browser = args[2];
+        const query = config.url.split('?')[1];
+        browser.url = `${browser.url}?${query}`;
+      }
+      return assign.apply(this, args);
+    };
+    // Something in Karma deps actually asserts the sub-keys of Object.assign,
+    // so make sure to copy those over too:
+    assign.call(Object, newAssign, assign);
+    Object.assign = newAssign;
```
Comment (elalish) [inline packages/3dom/karma.conf.js]: I love JavaScript! 😈

File: packages/model-viewer/karma.conf.js [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        os: 'iOS',
+        os_version: '12',
+        device: 'iPhone 8',
+        browser: 'iPhone',
+        real_mobile: 'true',
+        // BrowserStack seems to drop the port when redirecting to this special
+        // domain so we go there directly instead:
+        url: 'http://bs-local.com:9876'
+      },
+      'iOS Safari (iOS 12)': {
```
Comment (elalish) [inline packages/model-viewer/karma.conf.js]: Should we add an android device someday? Not saying today is that day...

File: packages/model-viewer/package.json [hunk 6]
```diff
@@ -76,35 +73,34 @@
...
     "@types/pngjs": "^3.4.0",
     "@ungap/event-target": "^0.1.0",
     "@webcomponents/webcomponentsjs": "~2.1.3",
     "chai": "^4.1.2",
-    "chokidar-cli": "^1.2.1",
-    "concurrently": "^4.0.1",
-    "cross-os": "^1.3.0",
     "focus-visible": "^5.0.1",
     "fullscreen-polyfill": "^1.0.2",
+    "http-server": "^0.12.1",
```
Comment (elalish) [inline packages/model-viewer/package.json]: Better than local-web-server?

File: packages/model-viewer/package.json [hunk 7]
```diff
@@ -76,35 +73,34 @@
...
     "intersection-observer": "^0.5.1",
     "lit-html": "^1.1.2",
-    "local-web-server": "^2.6.0",
-    "make-dir": "^1.3.0",
     "mocha": "^5.2.0",
```
Comment (elalish) [inline packages/model-viewer/package.json]: Does karma-mocha not include mocha?

File: packages/model-viewer/scripts/ci-run-tests.sh [hunk 8]
```diff
@@ -20,11 +20,8 @@ set -x
 
 if [ "${TEST_TYPE}" = "unit" ]; then
 
-  npm run test
+  npm run wct-local-chrome
 
-  if  [[ "${TRAVIS_PULL_REQUEST}" = "false" && "${TRAVIS_BRANCH}" != "master" ]]; then
```
Comment (elalish) [inline packages/model-viewer/scripts/ci-run-tests.sh]: Does this mean third parties start browserStack tests, or did this logic end up elsewhere?

File: packages/model-viewer/src/test/decorators-spec.ts [hunk 9]
```diff
@@ -50,7 +50,7 @@ suite('decorators', () => {
       customElements.define(tagName, class extends StyleableElement {});
 
       element = document.createElement(tagName) as StyleableElement;
-      document.body.appendChild(element);
+      document.body.insertBefore(element, document.body.firstChild);
```
Comment (elalish) [inline packages/model-viewer/src/test/decorators-spec.ts]: What's all this about? Something browser-specific?

File: packages/model-viewer/src/test/model-viewer-base-spec.ts [hunk 10]
```diff
@@ -338,10 +338,10 @@ suite('ModelViewerElementBase', () => {
 
       test('sets a model within viewport to be visible', async () => {
         await until(() => {
-          return elements[0][$scene].visible;
+          return elements[2][$scene].visible;
```
Comment (elalish) [inline packages/model-viewer/src/test/model-viewer-base-spec.ts]: Why this change?

File: packages/model-viewer/karma.conf.js (hunk 5 above)
Comment (cdata) [reply packages/model-viewer/karma.conf.js]: Yes, let's make it another day though. Tracked in #691

File: packages/model-viewer/package.json (hunk 6 above)
Comment (cdata) [reply packages/model-viewer/package.json]: Yes, and consistent with the other packages

File: packages/model-viewer/package.json (hunk 7 above)
Comment (cdata) [reply packages/model-viewer/package.json]: It does not, it manages the integration between the two.

File: packages/model-viewer/scripts/ci-run-tests.sh (hunk 8 above)
Comment (cdata) [reply packages/model-viewer/scripts/ci-run-tests.sh]: This got moved elsewhere

File: packages/model-viewer/src/test/decorators-spec.ts (hunk 9 above)
Comment (cdata) [reply packages/model-viewer/src/test/decorators-spec.ts]: Now that we are running in vanilla mocha, without additional intervention all of our elements get appended after the Mocha test results. This ensures that our test elements are inserted before those Mocha DOM elements.

File: packages/model-viewer/src/test/model-viewer-base-spec.ts (hunk 10 above)
Comment (cdata) [reply packages/model-viewer/src/test/model-viewer-base-spec.ts]: Because we reversed the order that the elements are inserted to the document.

**Comment by cdata** [comment]
I'm happy to share that "full" unit test runs (including Chrome, 2 x Firefox, 2 x Desktop Safari, 2 x MS Edge, 2 x iOS Safari, and IE11 across both @google/model-viewer and @google/3dom) are finishing in ~8 minutes. Unit tests and fidelity tests both complete in under 10 minutes. This is a 55-65% reduction in test time compared to our Travis/WCT/Sauce builds ([example build](https://travis-ci.org/github/google/model-viewer/builds/661719760)).

File: packages/model-viewer/scripts/ci-run-tests.sh [hunk 11]
```diff
@@ -20,11 +20,8 @@ set -x
 
 if [ "${TEST_TYPE}" = "unit" ]; then
 
-  npm run test
+  npm run wct-local-chrome
```
Comment (elalish) [inline packages/model-viewer/scripts/ci-run-tests.sh]: Still confused by this name; are we really still using WCT?

Comment (elalish) [reply packages/model-viewer/scripts/ci-run-tests.sh]: Oh wait, we don't actually need this file anymore, right? Since we're not using Travis anymore?

Comment (cdata) [reply packages/model-viewer/scripts/ci-run-tests.sh]: Good catch, this script is vestigial and has been removed 👍
